File: src/service/ServiceRegistry.js
const Logger = require('../monitor/Logger');

/**
 * ServiceRegistry - 服务注册表
 * 管理服务的注册和发现
 */
class ServiceRegistry {
  constructor() {
    this.logger = new Logger();
    this.services = new Map();
    this.subscribers = new Set();
  }
  
  /**
   * 注册服务
   */
  register(serviceId, service) {
    this.services.set(serviceId, service);
    this.notifySubscribers('service:registered', { serviceId, service });
    this.logger.info({ message: 'Service registered', serviceId });
  }
  
  /**
   * 注销服务
   */
  unregister(serviceId) {
    this.services.delete(serviceId);
    this.notifySubscribers('service:unregistered', { serviceId });
    this.logger.info({ message: 'Service unregistered', serviceId });
  }
  
  /**
   * 获取服务
   */
  get(serviceId) {
    return this.services.get(serviceId);
  }
  
  /**
   * 查询服务
   */
  query(criteria) {
    const results = [];
    for (const [id, service] of this.services.entries()) {
      if (this.matches(service, criteria)) {
        results.push(service);
      }
    }
    return results;
  }
  
  /**
   * 检查服务是否匹配条件
   */
  matches(service, criteria) {
    if (criteria.type && service.type !== criteria.type) return false;
    if (criteria.status && service.status !== criteria.status) return false;
    if (criteria.name && !service.name.includes(criteria.name)) return false;
    return true;
  }
  
  /**
   * 获取所有服务
   */
  getAll() {
    return Array.from(this.services.values());
  }
  
  /**
   * 订阅服务事件
   */
  subscribe(subscriber) {
    this.subscribers.add(subscriber);
  }
  
  /**
   * 取消订阅
   */
  unsubscribe(subscriber) {
    this.subscribers.delete(subscriber);
  }
  
  /**
   * 通知订阅者
   */
  notifySubscribers(event, data) {
    for (const subscriber of this.subscribers) {
      if (typeof subscriber === 'function') {
        subscriber(event, data);
      }
    }
  }
  
  /**
   * 获取统计信息
   */
  getStats() {
    const stats = {
      total: this.services.size,
      byStatus: {},
      byType: {}
    };
    
    for (const service of this.services.values()) {
      stats.byStatus[service.status] = (stats.byStatus[service.status] || 0) + 1;
      stats.byType[service.type] = (stats.byType[service.type] || 0) + 1;
    }
    
    return stats;
  }
}

module.exports = ServiceRegistry;

File: src/main/java/com/mapengine/service/ServiceRegistry.java
package com.mapengine.service;

import lombok.extern.slf4j.Slf4j;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;
import java.util.stream.Collectors;

@Slf4j
public class ServiceRegistry {
    private final Map<String, Service> services = new java.util.concurrent.ConcurrentHashMap<>();
    private final Set<java.util.function.BiConsumer<String, Service>> subscribers = new CopyOnWriteArraySet<>();

    public void register(String serviceId, Service service) {
        services.put(serviceId, service);
        notifySubscribers("service:registered", service);
        log.info("Service registered: {}", serviceId);
    }

    public void unregister(String serviceId) {
        services.remove(serviceId);
        notifySubscribers("service:unregistered", null);
        log.info("Service unregistered: {}", serviceId);
    }

    public Service get(String serviceId) { return services.get(serviceId); }

    public java.util.List<Service> query(Map<String, Object> criteria) {
        return services.values().stream().filter(s -> matches(s, criteria)).collect(Collectors.toList());
    }

    public boolean matches(Service s, Map<String, Object> criteria) {
        if (criteria == null || criteria.isEmpty()) return true;
        if (criteria.containsKey("type") && !criteria.get("type").equals(s.getType())) return false;
        if (criteria.containsKey("status") && !criteria.get("status").equals(s.getStatus())) return false;
        if (criteria.containsKey("name") && !s.getName().contains(String.valueOf(criteria.get("name")))) return false;
        return true;
    }

    public java.util.List<Service> getAll() { return new java.util.ArrayList<>(services.values()); }

    public void subscribe(java.util.function.BiConsumer<String, Service> subscriber) { subscribers.add(subscriber); }

    public void unsubscribe(java.util.function.BiConsumer<String, Service> subscriber) { subscribers.remove(subscriber); }

    private void notifySubscribers(String event, Service service) {
        for (var sub : subscribers) {
            try { sub.accept(event, service); } catch (Exception e) { log.warn("Subscriber threw", e); }
        }
    }

    public java.util.Map<String, Object> getStats() {
        java.util.Map<String, Object> stats = new java.util.HashMap<>();
        stats.put("total", services.size());
        return stats;
    }
}
